File: src/Component/Sensor.cpp
// *****************************************************************************
// *****************************************************************************
// Sensor.cpp
//
// Author: Jason Tost
// Date:   10.21.2016
//
// *****************************************************************************
// *****************************************************************************

#include "Sensor.hpp"

Sensor::Sensor(
   const String& id) :
      Component(id),
      pollTimer(0)
{
}

Sensor::Sensor(
   MessagePtr message) :
      Component(message),
      pollTimer(0)
{
}

Sensor::~Sensor()
{
}

// **************************************************************************
//                         MessageHandler interface

void Sensor::handleMessage(
   MessagePtr message)
{
   // read
   if (message->getMessageId() == "read")
   {
      Logger::logDebug(F("Sensor::handleMessage: read()"));

      sendReading(message->getSource());
   }
   // poll
   if (message->getMessageId() == "poll")
   {
      int period = message->getInt("period");

      Logger::logDebug(F("Sensor::handleMessage: poll(%d)"), period);

      poll(period);
   }
   // killSwitch
   else if (message->getMessageId() == "killSwitch")
   {
      Logger::logDebug(F("Sensor::handleMessage: killSwitch"));

      poll(0);
   }
   else
   {
      Component::handleMessage(message);
   }

   Messaging::freeMessage(message);
}

// **************************************************************************
//                         TimerListner interface

void Sensor::timeout(
   Timer* timer)
{
   onPoll();
}

// **************************************************************************

void Sensor::onPoll()
{
   broadcastReading();
}

void Sensor::poll(
   const int& period)
{
   if (pollTimer)
   {
      Timer::freeTimer(pollTimer);
   }

   if (period > 0)
   {
      pollTimer = Timer::newTimer("pollTimer", period, Timer::PERIODIC, this);

      if (pollTimer)
      {
         pollTimer->start();
      }
   }
}

void Sensor::sendReading(
      const String& destination)
{
   MessagePtr message = Messaging::newMessage();

   if (message)
   {
      message->setMessageId("sensorReading");
      message->setSource(getId());
      message->set("value", read());

      Messaging::send(message);
   }
}

void Sensor::broadcastReading()
{
   MessagePtr message = Messaging::newMessage();

   if (message)
   {
      message->setTopic(getId());
      message->setMessageId("sensorReading");
      message->setSource(getId());
      message->set("value", read());

      Messaging::publish(message);
   }
}

File: robox.js/component/sensor.js
function Sensor(id)
{
   this.id = id;
   
   this.parent = null;
   
   Sensor.prototype.ping = function()
   {
      var message = new Message();
      message.messageId = "ping";
      message.destination = this.id;

      parent.sendMessage(message);
   }
   
   Sensor.prototype.read = function()
   {
      var message = new Message();
      message.messageId = "read";
      message.destination = this.id;
      
      this.parent.sendMessage(message);
   }
   
   Sensor.prototype.poll = function(period)
   {
      var message = new Message();
      message.messageId = "poll";
      message.destination = this.id;
      message.period = period;
      
      this.parent.sendMessage(message);
   }
   
   Sensor.prototype.setUnits = function(units)
   {
      var message = new Message();
      message.messageId = "setUnits";
      message.destination = this.id;
      message.units = units;
      
      this.parent.sendMessage(message);
   }
}